File: src/types/knowledge.ts
/**
 * 知识库类型定义
 * 用于知识库管理和引用功能
 */

// ============================================================================
// 基础类型定义
// ============================================================================

/**
 * 知识库分类枚举值
 */
export type KnowledgeCategory = '技术文档' | '最佳实践' | '问题解决方案' | '培训资料' | '其他';

/**
 * 知识库状态枚举值
 */
export type KnowledgeStatus = '草稿' | '已发布' | '已归档';

// ============================================================================
// Element Plus Tag 类型映射
// ============================================================================

/**
 * Element Plus Tag 组件的类型
 * 用于不同状态/分类的视觉区分
 */
export type TagType = 'success' | 'warning' | 'danger' | 'info' | 'primary' | '';

// ============================================================================
// 选项常量接口
// ============================================================================

/**
 * 下拉选项接口定义
 * 用于表单中的选择器组件
 */
export interface OptionItem {
  /** 显示标签 */
  label: string;
  /** 实际值 */
  value: string;
  /** Element Plus Tag 类型 */
  type: TagType;
}

// ============================================================================
// 核心数据模型
// ============================================================================

/**
 * 知识库条目完整接口
 * 包含知识库的所有字段
 */
export interface Knowledge {
  /** 知识库 ID */
  id: string;
  
  /** 标题 */
  title: string;
  
  /** 内容摘要 */
  summary: string;
  
  /** 详细内容 */
  content: string;
  
  /** 分类 */
  category: KnowledgeCategory;
  
  /** 状态 */
  status: KnowledgeStatus;
  
  /** 作者 */
  author: string;
  
  /** 创建时间，ISO 8601 格式 */
  created_at: string;
  
  /** 最后更新时间，ISO 8601 格式 */
  updated_at: string;
  
  /** 发布时间，ISO 8601 格式 */
  published_at?: string;
  
  /** 标签列表 */
  tags: string[];
  
  /** 是否优质内容 */
  is_featured: boolean;
  
  /** 浏览次数 */
  view_count: number;
  
  /** 引用次数 */
  reference_count: number;
  
  /** 关联问题单号（可选） */
  related_issue_id?: string;
  
  /** 关联任务单号（可选） */
  related_task_id?: string;
}

// ============================================================================
// 请求接口定义
// ============================================================================

/**
 * 搜索知识库请求接口
 */
export interface SearchKnowledgeRequest {
  /** 页码，从 1 开始 */
  page?: number;
  
  /** 每页数量 */
  page_size?: number;
  
  /** 关键字搜索（标题/内容/标签） */
  keyword?: string;
  
  /** 分类筛选 */
  category?: KnowledgeCategory;
  
  /** 状态筛选 */
  status?: KnowledgeStatus;
  
  /** 作者筛选 */
  author?: string;
  
  /** 标签筛选 */
  tag?: string;
  
  /** 仅查询优质内容 */
  is_featured?: boolean;
}

/**
 * 知识库查询参数接口
 * 用于列表分页和筛选
 */
export interface KnowledgeQueryParams {
  /** 页码，从 1 开始 */
  page?: number;
  
  /** 每页数量 */
  page_size?: number;
  
  /** 关键字搜索（标题/内容/标签） */
  keyword?: string;
  
  /** 分类筛选 */
  category?: KnowledgeCategory;
  
  /** 状态筛选 */
  status?: KnowledgeStatus;
  
  /** 作者筛选 */
  author?: string;
  
  /** 标签筛选 */
  tag?: string;
  
  /** 仅查询优质内容 */
  is_featured?: boolean;
  
  /** 创建时间起始 */
  created_at_from?: string;
  
  /** 创建时间结束 */
  created_at_to?: string;
}

/**
 * 创建知识库请求接口
 */
export interface CreateKnowledgeRequest {
  /** 标题 */
  title: string;
  
  /** 内容摘要 */
  summary: string;
  
  /** 详细内容 */
  content: string;
  
  /** 分类 */
  category: KnowledgeCategory;
  
  /** 标签列表 */
  tags?: string[];
  
  /** 关联问题单号（可选） */
  related_issue_id?: string;
  
  /** 关联任务单号（可选） */
  related_task_id?: string;
}

/**
 * 更新知识库请求接口
 * 所有字段可选，支持部分更新
 */
export interface UpdateKnowledgeRequest {
  /** 标题 */
  title?: string;
  
  /** 内容摘要 */
  summary?: string;
  
  /** 详细内容 */
  content?: string;
  
  /** 分类 */
  category?: KnowledgeCategory;
  
  /** 标签列表 */
  tags?: string[];
  
  /** 关联问题单号 */
  related_issue_id?: string;
  
  /** 关联任务单号 */
  related_task_id?: string;
}

/**
 * 标记优质内容请求接口
 */
export interface MarkAsFeaturedRequest {
  /** 是否优质 */
  is_featured: boolean;
  
  /** 标记说明（可选） */
  note?: string;
}

/**
 * 记录引用请求接口
 */
export interface RecordReferenceRequest {
  /** 引用来源描述 */
  source?: string;
  
  /** 引用说明 */
  note?: string;
}

// ============================================================================
// 响应接口定义
// ============================================================================

/**
 * 知识库列表响应接口
 */
export interface KnowledgeListResponse {
  /** 知识库列表 */
  items: Knowledge[];
  
  /** 总数量 */
  total: number;
  
  /** 当前页码 */
  page: number;
  
  /** 每页数量 */
  page_size: number;
  
  /** 总页数 */
  total_pages: number;
}

/**
 * 知识库详情响应接口
 */
export interface KnowledgeDetailResponse {
  /** 知识库数据 */
  data: Knowledge;
}

/**
 * 通用消息响应接口
 */
export interface MessageResponse {
  /** 消息内容 */
  message: string;
  
  /** 相关 ID（可选） */
  id?: string;
}

// ============================================================================
// 选项常量数组
// ============================================================================

/**
 * 分类选项常量数组
 */
export const categoryOptions: OptionItem[] = [
  { label: '技术文档', value: '技术文档', type: 'primary' },
  { label: '最佳实践', value: '最佳实践', type: 'success' },
  { label: '问题解决方案', value: '问题解决方案', type: 'warning' },
  { label: '培训资料', value: '培训资料', type: 'info' },
  { label: '其他', value: '其他', type: '' },
];

/**
 * 状态选项常量数组
 */
export const statusOptions: OptionItem[] = [
  { label: '草稿', value: '草稿', type: 'info' },
  { label: '已发布', value: '已发布', type: 'success' },
  { label: '已归档', value: '已归档', type: '' },
];

// ============================================================================
// 工具函数
// ============================================================================

/**
 * 获取分类标签类型
 * @param category 知识库分类
 * @returns Element Plus Tag 类型
 */
export function getCategoryTagType(category: KnowledgeCategory): TagType {
  const option = categoryOptions.find(opt => opt.value === category);
  return option?.type || '';
}

/**
 * 获取状态标签类型
 * @param status 知识库状态
 * @returns Element Plus Tag 类型
 */
export function getStatusTagType(status: KnowledgeStatus): TagType {
  const option = statusOptions.find(opt => opt.value === status);
  return option?.type || '';
}

File: backend/schemas/knowledge.py
"""
Knowledge 模块 Pydantic Schema 定义
AutoCraft 知识管理系统 - Pydantic v2 数据验证模型
"""
from datetime import datetime
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


# ============================================================================
# 枚举定义
# ============================================================================

class KnowledgeType(str, Enum):
    """知识类型枚举"""
    DOCUMENT = "文档"
    TUTORIAL = "教程"
    BEST_PRACTICE = "最佳实践"
    CASE_STUDY = "案例研究"
    FAQ = "常见问题"
    REFERENCE = "参考资料"


class KnowledgeStatus(str, Enum):
    """知识状态枚举"""
    DRAFT = "草稿"
    PUBLISHED = "已发布"
    ARCHIVED = "已归档"


class KnowledgeQuality(str, Enum):
    """知识质量等级枚举"""
    NORMAL = "普通"
    FEATURED = "优质"


# ============================================================================
# 基础模型
# ============================================================================

class KnowledgeBase(BaseModel):
    """Knowledge 基础模型"""
    title: str = Field(..., max_length=200, description="知识标题")
    summary: str = Field(..., description="知识摘要")
    content: str = Field(..., description="知识内容")
    knowledge_type: KnowledgeType = Field(..., description="知识类型")
    tags: Optional[List[str]] = Field(default_factory=list, description="标签列表")


# ============================================================================
# 请求模型
# ============================================================================

class KnowledgeCreate(KnowledgeBase):
    """创建 Knowledge 请求模型"""
    project_id: Optional[int] = Field(None, description="项目 ID")
    category_id: Optional[int] = Field(None, description="分类 ID")
    created_by: str = Field(..., description="创建人")


class KnowledgeUpdate(BaseModel):
    """更新 Knowledge 请求模型（所有字段可选）"""
    title: Optional[str] = Field(None, max_length=200, description="知识标题")
    summary: Optional[str] = Field(None, description="知识摘要")
    content: Optional[str] = Field(None, description="知识内容")
    knowledge_type: Optional[KnowledgeType] = Field(None, description="知识类型")
    status: Optional[KnowledgeStatus] = Field(None, description="状态")
    tags: Optional[List[str]] = Field(None, description="标签列表")
    category_id: Optional[int] = Field(None, description="分类 ID")


class KnowledgeSearchFilter(BaseModel):
    """Knowledge 搜索条件模型"""
    page: int = Field(default=1, ge=1, le=1000, description="页码")
    page_size: int = Field(default=10, ge=1, le=100, description="每页数量")
    keyword: Optional[str] = Field(None, description="关键字搜索（标题/摘要/内容）")
    knowledge_type: Optional[KnowledgeType] = Field(None, description="知识类型筛选")
    status: Optional[KnowledgeStatus] = Field(None, description="状态筛选")
    quality: Optional[KnowledgeQuality] = Field(None, description="质量等级筛选")
    category_id: Optional[int] = Field(None, description="分类 ID 筛选")
    project_id: Optional[int] = Field(None, description="项目 ID 筛选")
    tags: Optional[List[str]] = Field(None, description="标签筛选")
    created_by: Optional[str] = Field(None, description="创建人筛选")
    is_featured: Optional[bool] = Field(None, description="是否优质筛选")
    created_at_from: Optional[datetime] = Field(None, description="创建时间起始")
    created_at_to: Optional[datetime] = Field(None, description="创建时间结束")


class KnowledgeBaseCreate(BaseModel):
    """创建 KnowledgeBase 请求模型"""
    issue_id: int = Field(..., description="关联的 Issue ID")
    issue_no: str = Field(..., max_length=50, description="问题编号")
    title: str = Field(..., max_length=200, description="知识条目标题")
    description: Optional[str] = Field(None, description="问题描述")
    category: str = Field(..., max_length=50, description="知识分类")
    solution: Optional[str] = Field(None, description="解决方案")
    solution_process: Optional[str] = Field(None, description="解决过程")
    root_cause: Optional[str] = Field(None, description="根本原因")
    solution_approach: Optional[str] = Field(None, description="解决方法")
    lessons_learned: Optional[str] = Field(None, description="经验总结")
    prevention_measures: Optional[str] = Field(None, description="预防措施")
    is_featured: Optional[bool] = Field(default=False, description="是否精选")


class KnowledgeBaseUpdate(BaseModel):
    """更新 KnowledgeBase 请求模型（所有字段可选）"""
    issue_id: Optional[int] = Field(None, description="关联的 Issue ID")
    issue_no: Optional[str] = Field(None, max_length=50, description="问题编号")
    title: Optional[str] = Field(None, max_length=200, description="知识条目标题")
    description: Optional[str] = Field(None, description="问题描述")
    category: Optional[str] = Field(None, max_length=50, description="知识分类")
    solution: Optional[str] = Field(None, description="解决方案")
    solution_process: Optional[str] = Field(None, description="解决过程")
    root_cause: Optional[str] = Field(None, description="根本原因")
    solution_approach: Optional[str] = Field(None, description="解决方法")
    lessons_learned: Optional[str] = Field(None, description="经验总结")
    prevention_measures: Optional[str] = Field(None, description="预防措施")
    is_featured: Optional[bool] = Field(None, description="是否精选")


class KnowledgeBaseSearchFilter(BaseModel):
    """KnowledgeBase 搜索条件模型"""
    page: int = Field(default=1, ge=1, le=1000, description="页码")
    page_size: int = Field(default=10, ge=1, le=100, description="每页数量")
    keyword: Optional[str] = Field(None, description="关键字搜索（标题/描述/解决方案等）")
    category: Optional[str] = Field(None, description="分类筛选")
    is_featured: Optional[bool] = Field(None, description="是否精选筛选")


class KnowledgeBaseResponse(BaseModel):
    """KnowledgeBase 响应模型"""
    id: int = Field(..., description="知识 ID")
    issue_id: int = Field(..., description="关联的 Issue ID")
    issue_no: str = Field(..., max_length=50, description="问题编号")
    title: str = Field(..., max_length=200, description="知识条目标题")
    description: Optional[str] = Field(None, description="问题描述")
    category: str = Field(..., max_length=50, description="知识分类")
    solution: Optional[str] = Field(None, description="解决方案")
    solution_process: Optional[str] = Field(None, description="解决过程")
    root_cause: Optional[str] = Field(None, description="根本原因")
    solution_approach: Optional[str] = Field(None, description="解决方法")
    lessons_learned: Optional[str] = Field(None, description="经验总结")
    prevention_measures: Optional[str] = Field(None, description="预防措施")
    is_featured: bool = Field(default=False, description="是否精选")
    view_count: int = Field(default=0, description="浏览次数")
    reference_count: int = Field(default=0, description="引用次数")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    model_config = ConfigDict(from_attributes=True)


class KnowledgeBaseListResponse(BaseModel):
    """KnowledgeBase 列表响应模型（分页）"""
    total: int = Field(..., description="总记录数")
    items: List[KnowledgeBaseResponse] = Field(..., description="知识库列表")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")
    total_pages: int = Field(..., description="总页数")


class KnowledgeFeatureRequest(BaseModel):
    """标记优质 Knowledge 请求模型"""
    is_featured: bool = Field(..., description="是否标记为优质")


class KnowledgeReferenceRequest(BaseModel):
    """引用 Knowledge 请求模型"""
    reference_context: str = Field(..., description="引用上下文")
    referenced_by: str = Field(..., description="引用操作人/资源标识")


# ============================================================================
# 响应模型
# ============================================================================

class KnowledgeResponse(KnowledgeBase):
    """Knowledge 响应模型"""
    id: int = Field(..., description="Knowledge ID")
    knowledge_no: str = Field(..., description="知识编号")
    status: KnowledgeStatus = Field(..., description="状态")
    quality: KnowledgeQuality = Field(default=KnowledgeQuality.NORMAL, description="质量等级")
    is_featured: bool = Field(default=False, description="是否优质")
    featured_reason: Optional[str] = Field(None, description="标记优质理由")
    featured_at: Optional[datetime] = Field(None, description="标记优质时间")
    featured_by: Optional[str] = Field(None, description="标记优质操作人")
    project_id: Optional[int] = Field(None, description="项目 ID")
    category_id: Optional[int] = Field(None, description="分类 ID")
    category_name: Optional[str] = Field(None, description="分类名称")
    view_count: int = Field(default=0, description="浏览次数")
    like_count: int = Field(default=0, description="点赞次数")
    reference_count: int = Field(default=0, description="引用次数")
    created_by: str = Field(..., description="创建人")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    published_at: Optional[datetime] = Field(None, description="发布时间")
    archived_at: Optional[datetime] = Field(None, description="归档时间")

    model_config = ConfigDict(from_attributes=True)


class KnowledgeListResponse(BaseModel):
    """Knowledge 列表响应模型（分页）"""
    total: int = Field(..., description="总记录数")
    items: List[KnowledgeResponse] = Field(..., description="Knowledge 列表")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")
    total_pages: int = Field(..., description="总页数")


class KnowledgeDetailResponse(BaseModel):
    """Knowledge 详情响应模型"""
    data: KnowledgeResponse = Field(..., description="Knowledge 详情")


# ============================================================================
# 通用响应模型
# ============================================================================

class MessageResponse(BaseModel):
    """通用消息响应模型"""
    message: str = Field(..., description="消息内容")
    knowledge_no: Optional[str] = Field(None, description="知识编号")
